File: src/MediaClassification.cpp
#include "MediaClassification.h"

#include <QFileInfo>
#include <QMimeDatabase>
#include <QSet>

namespace {
const QSet<QString> &imageExtensions()
{
    static const QSet<QString> value = {
        QStringLiteral("jpg"), QStringLiteral("jpeg"), QStringLiteral("png"), QStringLiteral("gif"), QStringLiteral("webp"),
        QStringLiteral("avif"), QStringLiteral("heic"), QStringLiteral("heif"), QStringLiteral("jxl"), QStringLiteral("bmp"),
        QStringLiteral("tif"), QStringLiteral("tiff"), QStringLiteral("ico"), QStringLiteral("icns"),
    };
    return value;
}

const QSet<QString> &videoExtensions()
{
    static const QSet<QString> value = {
        QStringLiteral("mp4"), QStringLiteral("m4v"), QStringLiteral("mov"), QStringLiteral("webm"), QStringLiteral("mkv"),
        QStringLiteral("avi"), QStringLiteral("mpg"), QStringLiteral("mpeg"), QStringLiteral("ts"), QStringLiteral("mts"),
        QStringLiteral("m2ts"), QStringLiteral("flv"), QStringLiteral("ogv"), QStringLiteral("3gp"), QStringLiteral("3g2"),
        QStringLiteral("wmv"), QStringLiteral("asf"), QStringLiteral("mxf"),
    };
    return value;
}

const QSet<QString> &programExtensions()
{
    static const QSet<QString> value = {
        QStringLiteral("dmg"), QStringLiteral("pkg"), QStringLiteral("app"), QStringLiteral("ipa"), QStringLiteral("apk"),
        QStringLiteral("exe"), QStringLiteral("msi"), QStringLiteral("deb"), QStringLiteral("rpm"), QStringLiteral("appimage"),
        QStringLiteral("jar"), QStringLiteral("bin"), QStringLiteral("run"), QStringLiteral("command"), QStringLiteral("bat"),
        QStringLiteral("ps1"),
    };
    return value;
}

const QSet<QString> &archiveExtensions()
{
    static const QSet<QString> value = {
        QStringLiteral("zip"), QStringLiteral("tar"), QStringLiteral("gz"), QStringLiteral("tgz"), QStringLiteral("bz2"),
        QStringLiteral("xz"), QStringLiteral("7z"), QStringLiteral("rar"), QStringLiteral("zst"),
    };
    return value;
}

const QSet<QString> &documentExtensions()
{
    static const QSet<QString> value = {
        QStringLiteral("pdf"), QStringLiteral("txt"), QStringLiteral("md"), QStringLiteral("rtf"), QStringLiteral("doc"),
        QStringLiteral("docx"), QStringLiteral("xls"), QStringLiteral("xlsx"), QStringLiteral("ppt"), QStringLiteral("pptx"),
        QStringLiteral("odt"), QStringLiteral("ods"), QStringLiteral("odp"), QStringLiteral("csv"), QStringLiteral("json"),
        QStringLiteral("xml"), QStringLiteral("yaml"), QStringLiteral("yml"),
    };
    return value;
}

QString normalizedMimeType(const QString &mimeType)
{
    const int separator = mimeType.indexOf(QLatin1Char(';'));
    return (separator >= 0 ? mimeType.left(separator) : mimeType).trimmed().toLower();
}
}

MediaCategory MediaClassification::category(const QString &filename, const QString &mimeType)
{
    const QString extension = filenameExtension(filename);
    if (!extension.isEmpty()) {
        if (imageExtensions().contains(extension)) {
            return MediaCategory::Images;
        }
        if (videoExtensions().contains(extension)) {
            return MediaCategory::Videos;
        }
        if (programExtensions().contains(extension)) {
            return MediaCategory::Programs;
        }
        if (archiveExtensions().contains(extension)) {
            return MediaCategory::Archives;
        }
        if (documentExtensions().contains(extension)) {
            return MediaCategory::Documents;
        }

        QMimeDatabase database;
        const QMimeType type = database.mimeTypeForFile(QStringLiteral("file.") + extension, QMimeDatabase::MatchExtension);
        if (type.name().startsWith(QStringLiteral("image/"))) {
            return MediaCategory::Images;
        }
        if (type.name().startsWith(QStringLiteral("video/"))) {
            return MediaCategory::Videos;
        }
        if (type.name().startsWith(QStringLiteral("audio/"))) {
            return MediaCategory::Audio;
        }
    }

    const QString normalized = normalizedMimeType(mimeType);
    if (normalized.startsWith(QStringLiteral("image/"))) {
        return MediaCategory::Images;
    }
    if (normalized.startsWith(QStringLiteral("video/"))) {
        return MediaCategory::Videos;
    }
    if (normalized.startsWith(QStringLiteral("audio/"))) {
        return MediaCategory::Audio;
    }
    if (normalized == QStringLiteral("application/webm") || normalized == QStringLiteral("application/x-matroska")) {
        return MediaCategory::Videos;
    }
    if (normalized == QStringLiteral("application/pdf") || normalized.startsWith(QStringLiteral("text/"))) {
        return MediaCategory::Documents;
    }
    if (normalized.contains(QStringLiteral("zip")) || normalized.contains(QStringLiteral("compressed")) || normalized.contains(QStringLiteral("archive"))) {
        return MediaCategory::Archives;
    }
    if (normalized.contains(QStringLiteral("msi")) || normalized.contains(QStringLiteral("executable")) || normalized.contains(QStringLiteral("application/x-dosexec"))) {
        return MediaCategory::Programs;
    }

    return MediaCategory::Other;
}

QString MediaClassification::preferredExtension(const QString &filename, const QString &mimeType)
{
    const QString extension = filenameExtension(filename);
    if (!extension.isEmpty()) {
        return extension;
    }

    const QString normalized = normalizedMimeType(mimeType);
    if (normalized.isEmpty()) {
        return {};
    }

    QMimeDatabase database;
    const QMimeType type = database.mimeTypeForName(normalized);
    return type.preferredSuffix().toLower();
}

QString MediaClassification::filenameExtension(const QString &filename)
{
    if (filename.isEmpty()) {
        return {};
    }

    const QString extension = QFileInfo(filename).suffix().toLower();
    return extension;
}

File: backend/src/media_classification.rs
use mime_guess::MimeGuess;

use crate::domain::MediaCategory;

const IMAGE_EXTENSIONS: &[&str] = &[
    "jpg", "jpeg", "png", "gif", "webp", "avif", "heic", "heif", "jxl", "bmp", "tif", "tiff",
    "ico", "icns",
];
const VIDEO_EXTENSIONS: &[&str] = &[
    "mp4", "m4v", "mov", "webm", "mkv", "avi", "mpg", "mpeg", "ts", "mts", "m2ts", "flv", "ogv",
    "3gp", "3g2", "wmv", "asf", "mxf",
];
const PROGRAM_EXTENSIONS: &[&str] = &[
    "dmg", "pkg", "app", "ipa", "apk", "exe", "msi", "deb", "rpm", "appimage", "jar", "bin", "run",
    "command", "bat", "ps1",
];
const ARCHIVE_EXTENSIONS: &[&str] = &["zip", "tar", "gz", "tgz", "bz2", "xz", "7z", "rar", "zst"];
const DOCUMENT_EXTENSIONS: &[&str] = &[
    "pdf", "txt", "md", "rtf", "doc", "docx", "xls", "xlsx", "ppt", "pptx", "odt", "ods", "odp",
    "csv", "json", "xml", "yaml", "yml",
];

pub fn category(filename: Option<&str>, mime_type: Option<&str>) -> MediaCategory {
    if let Some(ext) = filename_extension(filename) {
        if IMAGE_EXTENSIONS.contains(&ext.as_str()) {
            return MediaCategory::Images;
        }
        if VIDEO_EXTENSIONS.contains(&ext.as_str()) {
            return MediaCategory::Videos;
        }
        if PROGRAM_EXTENSIONS.contains(&ext.as_str()) {
            return MediaCategory::Programs;
        }
        if ARCHIVE_EXTENSIONS.contains(&ext.as_str()) {
            return MediaCategory::Archives;
        }
        if DOCUMENT_EXTENSIONS.contains(&ext.as_str()) {
            return MediaCategory::Documents;
        }

        if let Some(mime) = MimeGuess::from_ext(&ext).first() {
            if mime.type_() == mime::IMAGE {
                return MediaCategory::Images;
            }
            if mime.type_() == mime::VIDEO {
                return MediaCategory::Videos;
            }
            if mime.type_() == mime::AUDIO {
                return MediaCategory::Audio;
            }
        }
    }

    if let Some(mime) = normalized_mime_type(mime_type) {
        if mime.starts_with("image/") {
            return MediaCategory::Images;
        }
        if mime.starts_with("video/") {
            return MediaCategory::Videos;
        }
        if mime.starts_with("audio/") {
            return MediaCategory::Audio;
        }
        if mime == "application/webm" || mime == "application/x-matroska" {
            return MediaCategory::Videos;
        }
        if mime == "application/pdf" || mime.starts_with("text/") {
            return MediaCategory::Documents;
        }
        if mime.contains("zip") || mime.contains("compressed") || mime.contains("archive") {
            return MediaCategory::Archives;
        }
        if mime.contains("msi")
            || mime.contains("executable")
            || mime.contains("application/x-dosexec")
        {
            return MediaCategory::Programs;
        }
    }

    MediaCategory::Other
}

pub fn preferred_extension(filename: Option<&str>, mime_type: Option<&str>) -> Option<String> {
    if let Some(ext) = filename_extension(filename) {
        return Some(ext);
    }

    let mime = normalized_mime_type(mime_type)?;
    match mime.as_str() {
        "image/jpeg" => Some("jpg".to_owned()),
        "image/png" => Some("png".to_owned()),
        "image/gif" => Some("gif".to_owned()),
        "image/webp" => Some("webp".to_owned()),
        "image/avif" => Some("avif".to_owned()),
        "image/heic" => Some("heic".to_owned()),
        "image/heif" => Some("heif".to_owned()),
        "image/jxl" => Some("jxl".to_owned()),
        "image/bmp" => Some("bmp".to_owned()),
        "image/tiff" => Some("tiff".to_owned()),
        "video/mp4" | "application/mp4" => Some("mp4".to_owned()),
        "video/quicktime" => Some("mov".to_owned()),
        "video/webm" | "application/webm" => Some("webm".to_owned()),
        "video/x-matroska" | "application/x-matroska" => Some("mkv".to_owned()),
        "video/x-msvideo" => Some("avi".to_owned()),
        "video/x-ms-wmv" => Some("wmv".to_owned()),
        "video/mpeg" => Some("mpeg".to_owned()),
        "video/3gpp" => Some("3gp".to_owned()),
        "video/3gpp2" => Some("3g2".to_owned()),
        _ => mime_guess::get_mime_extensions_str(&mime)
            .and_then(|extensions| extensions.first().copied())
            .map(str::to_lowercase),
    }
}

pub fn filename_extension(filename: Option<&str>) -> Option<String> {
    let filename = filename?;
    let ext = std::path::Path::new(filename)
        .extension()?
        .to_str()?
        .trim()
        .to_lowercase();
    if ext.is_empty() { None } else { Some(ext) }
}

fn normalized_mime_type(mime_type: Option<&str>) -> Option<String> {
    let mime_type = mime_type?;
    let normalized = mime_type.split(';').next()?.trim().to_lowercase();
    if normalized.is_empty() {
        None
    } else {
        Some(normalized)
    }
}
